### ISSUE TITLE
Support Multiple Document Epub

### ISSUE THREAD

**Issue by shoesforindustry** [typora/typora-issues#824]
As you can pass a series of separate files to process in Pandoc. It would be great if you could support a multiple markdown document epub, one made with a series of separate markdown files... it is the way I expect most writers work. 

So say we have some separate documents, all in the same folder.

+ introduction.md
+ chapter-1.md
+ chapter-2.md
+ notes.md
+ biography.md
+ contact.md

I was thinking it could be easy to add a YAML identifier likes 'files', which contained a comma separated list of files for the epub, in the order they should appear. You would probably only need the YAML in the introduction/first page of this document set.

For example in the `introduction.md` above include the YAML:

```
---
title: My Book Title
author: A.N. Author
cover-image: coverimage.jpg
files: chapter-1.md, chapter-2.md, notes.md, biography.md, contact.md
---

# Introduction

This is an introduction page which acts like a build page for your epub

```

Any thoughts?

**Comment by shoesforindustry** [comment]
Then when export/compile Epub in Typora (with Pandoc) you can pass the file array e.g.
`pandoc introduction.md chapter-1.md chapter-2.md notes.md biography.md contact.md -o book.epub`
Of course you could also add in setting for custom templates (--template ) and custom stylesheets (--epub-stylesheet)

**Comment by vassudanagunta** [comment]
I'm pretty sure this is a bad idea because it essentially turns Typora into a Static Site Generator. 

Typora should focus on being THE BEST Markdown editor and leave the publishing function to publishing tools such as Jekyll, VuePress, Hugo, Pandoc, etc.  You wouldn't ask the publishing tools to be a Markdown Editor, would you?

Instead, I would propose that Typora be able to execute an external publish command, e.g. one of the above publishing tools. 

I'm working on [a standard](https://github.com/Markaround/markaround) and [a fork of Hugo](https://github.com/Markaround/hugo-af) that to make authoring *assemblies* of Markdown pages easy, generic and portable. Star the repos to stay informed.

**Comment by EditorJack** [comment]
I agree that making Typora the best Markdown editor available is the primary goal (actually, it already is), but for those of us who are merely writers rather than software geeks, using something like Jekyll or VuePress isn't a good option. The request from shoesforindustry is an excellent idea, in my opinion. After all, why not? Typora already supports exporting through Pandoc; why not make it possible to add a list of documents and export  them en masse?

I'm also an editor, and I'm constantly recommending Typora to authors as an excellent alternative to the bloated Microsoft Word. But again, authors need an easy way to list their Markdown files (chapters or notes) in an arbitrary order and then export them as one document (a book).

Please, please add this functionality. Thank you for listing, and thank you for such a marvelous Markdown editor.

**Comment by vassudanagunta** [comment]
@EditorJack I am making a version of Hugo that is ultra author friendly. In fact, it's named *[Hugo AF](https://github.com/Markaround/hugo-af)* and the AF stands for [Author's First](https://github.com/Markaround/markaround/blob/placeholder/authors-first-manifesto.md). You use pure Markdown, organize it how you want, and run it through Hugo, which has a super fast native single file executable so there is nothing to install. 

Stay tuned by starring the repo.

**Comment by EditorJack** [comment]
Hi, Vas.

Thanks very much for this information. You wrote:

 > Stay tuned by starring the repo.

Done! I'm really excited about this project. Please keep me posted. And 
if I can be of help, please let me know.

Best wishes,
Jack Lyon
The Editorium
http://www.editorium.com

**Comment by rster2002** [comment]
I would be nice of typora had it's own exporter in the form of a cli (or something that could be programmed) because for example pandoc doesn't support the `[toc]` and will just export it as `<p>[toc]</p>`. I can see this kind of issue also happening to things the the sequence diagram or flowchart.

It's not a deal breaker not having this option, but it would help a lot. For example, I take my notes using typora, and I then export them to HTML so others can view them on my website (as a sort of low effort way of sharing my notes). But I do need to update multiple files each time I create a new note because I then have to update links etc.

**Comment by G3Kappa** [comment]
> 
> 
> I would be nice of typora had it's own exporter in the form of a cli (or something that could be programmed) because for example pandoc doesn't support the `[toc]` and will just export it as `<p>[toc]</p>`. I can see this kind of issue also happening to things the the sequence diagram or flowchart.
> 
> It's not a deal breaker not having this option, but it would help a lot. For example, I take my notes using typora, and I then export them to HTML so others can view them on my website (as a sort of low effort way of sharing my notes). But I do need to update multiple files each time I create a new note because I then have to update links etc.

+1

I use Typora for pretty much everything, but lately I've been taking notes at work and sharing them on Microsoft Teams. And Teams is rather disappointing when it comes to previewing Markdown files (well, not only that...), so I've been exporting my notes as HTML - which renders great! However, as @rster2002 said, updating notes is a chore now.

Furthermore, I don't get @vassudanagunta 's reasoning. Exporting an individual file is already an automatic process that requires no user interaction, so what's the problem with exporting an entire folder at once? If people need a static site generator, they'll find one, but I'm merely looking to export my notes independently from one another.

**Comment by redoxmatlab** [comment]
Hi, I am no expert and do not use typora in producing static websites. I use Typora for producing technical documents and having the ability to assemble all (or selected) documents in a folder to PDF or DOCX is very time saving and valuable. 

 @abnerlee Could you please implement this functionality soon.

**Comment by sahilrajput03** [comment]
Please bring this feature, I'm currently using [Markdown styles](https://github.com/mixu/markdown-styles) with some batch script to do my work. I would love typora more if you could ship this feature.
And that batch script is getting tedious to handle month by month.

**Comment by kodaman2** [comment]
> @EditorJack I am making a version of [Hugo](https://gohugo.io) that is ultra author friendly. In fact, it's named _[Hugo AF](https://github.com/Markaround/hugo-af)_ and the AF stands for [Author's First](https://github.com/Markaround/markaround/blob/placeholder/authors-first-manifesto.md). You use pure Markdown, organize it how you want, and run it through Hugo, which has a super fast native single file executable so there is nothing to install.
> 
> Stay tuned by starring the repo.

I'm a software engineer, and I don't even want to deal with pandoc/latex, etc, even though I can, and I have in the past. Plenty people just want to focus on writing N markdown files, and then hitting a print button.  This isn't a publishing feature, just think if Microsoft Word made their users have to learn an api or cli tool to print a document, they'll be looking elsewhere.

Btw I wrote a little script which takes chapters or md files, and prints to pdf. Maybe it will help someone in this thread.

https://github.com/kodaman2/markdown-book

**Comment by tom-milner** [comment]
I agree with @TheFern2, I shouldn't have to dive into a terminal just to print the university notes that are already open in front of me.
Being able to export all my notes into a chosen format at once would be a massive help, as would the ability to select multiple files at once in the file explorer.

**Comment by TheFern2** [comment]
> I agree with @TheFern2, I shouldn't have to dive into a terminal just to print the university notes that are already open in front of me.
> Being able to export all my notes into a chosen format at once would be a massive help, as would the ability to select multiple files at once in the file explorer.

I wrote a vscode extension, that simplifies the printing process. It can print a single document, or a whole book. Let me know if it helps. I mainly printed from markdown to pdf, but in the options you can change extensions so should be able to print anything.

https://marketplace.visualstudio.com/items?itemName=TheFern.vscode-gutenberg

**Comment by rster2002** [comment]
Does this extention support mermaid graphs and latex blocks?

**Comment by TheFern2** [comment]
> Does this extention support mermaid graphs and latex blocks?

I have no idea what mermaid graphs are haha. I did tried latex blocks with my markdown and it prints fine. Look on my repo under tests folders and look at book 1 in raw mode. Basically I have a preamble with a few packages, then a few math equations. But latex should work fine no matter what, the extension uses pandoc and latex (mikex) in my case on the backend. Give it a try and if you run into any issues let me know fernandobe+git@protonmail.com or open an issue on the repo.

**Comment by aurelien-opex** [comment]
#2948

**Comment by TheFern2** [comment]
@mikedeskevich see previous comment, my vscode plugin does exactly what you need. You can work on typora and when you want to print it out use my plugin. I welcome any feedback.

**Comment by TheFern2** [comment]
> How do we install that plugin in Typora?

My extension is a vscode plugin not typora (vscode is a text editor), so if you like typora for editing just write your document(s) there, and when you are ready to print a book or whatever switch over to vscode and use this plugin https://marketplace.visualstudio.com/items?itemName=TheFern.vscode-gutenberg there is a video there on usage and also on the README you have instructions on the requirements, you need pandoc and latex installed.

**Comment by TheFern2** [comment]
Just look at the date on these issues, Typora devs have never made an attempt to solve the issue of exporting an entire folder and is a paid app afaik, I've used it long time ago and I had to develop first a python script and then a vscode extension to solve this issue. Typora handles all export/import by individual files, and most likely it will never change.

My extension provides a solution to this issue and your issue, if you don't want to use it that's fine, wait on Typora devs then. Unsubscribing from this issue, best of luck.

**Comment by aloboa** [comment]
@TheFern2 Do not take it wrong. I am grateful for your suggestion, but cannot accept that the Typora devs close my question with that as it had been solved. No critics to you. My issue #5500 is very new and has been redirected here and closed.

**Comment by TheFern2** [comment]
I didn't take it wrong, to me it looks like #5500 is a duplicate of #824 as in taking an entire folder and making it pdf or html (both of which my extension can do). I don't think @abnerlee meant my extension is a solution. About 3 years ago I had the exact same issue when I was using typora where I was trying to print a book, so I made this script https://github.com/TheFern2/markdown-book then I made the extension. All I am saying is that typora devs may or may not implement this feature any time soon, maybe never, who knows, use the tools you have today or you'll be waiting forever.

**Comment by aloboa** [comment]
They can do whatever they want, of course, but not to close issues that have not been solved as if they had been solved. In particular for a paid software.

**Comment by Ziflin** [comment]
Is there anything being done on this 7 year old issue? It appears that multiple people have opened similar issues regarding the ability to export a folder as a single PDF/etc. and they all redirect here with no useful information or updates for Typora. Our team was looking into using Typora for our projects documentation, but without this feature we're unable to use it as cramming everything into a single .md is not possible.

**Comment by TheFern2** [comment]
> Is there anything being done on this 7 year old issue? It appears that multiple people have opened similar issues regarding the ability to export a folder as a single PDF/etc. and they all redirect here with no useful information or updates for Typora. Our team was looking into using Typora for our projects documentation, but without this feature we're unable to use it as cramming everything into a single .md is not possible.

Use my vscode extension as mentioned previously. Typora isn't doing anything extraordinary these are just markdown files. Use vscode or obsidian to write files, you can even use typora for editing just not printing lol. Then look for a tool to print the documents my plugin in vscode uses pandoc in the background you can use that, or use pandoc on your own. Why do people lock themselves to typora for printing I'll never understand that. 🤷 Think outside the box folks.

**Comment by mikedeskevich** [comment]
> Is there anything being done on this 7 year old issue? It appears that multiple people have opened similar issues regarding the ability to export a folder as a single PDF/etc. and they all redirect here with no useful information or updates for Typora. Our team was looking into using Typora for our projects documentation, but without this feature we're unable to use it as cramming everything into a single .md is not possible.

I love Typora too much for the live-preview and editing so what I've done is use the m4 macro processor (https://gnu.org/software/m4/manual/) to put together my "chapters" into one big md file and then I use Typora to open that and print/save it as the format I need.  So my workflow is 

1. write chapters or sections as .md files in Typora
2. create a .m4 file to explain how to combine them
3. use m4 to create a new concatenated md file
4. use Typora on the new .md file

The hassle is worth it to me. I also reuse chapters and sections in different reports, so many times all I need to do is create a new .m4 file and I'm 90% of the way done.

**Comment by shoesforindustry** [comment]
I don't think you have to do anything special to concatenate several txt/md files into one file for further processing? Most UNIX / macOS have the 'cat' command e.g.

cat file1.md file2.md file3.md > file4.md

RussOn 3 Aug 2024, at 18:55, 

Doug ***@***.***> wrote:﻿
Is there anything being done on this 7 year old issue? It appears that multiple people have opened similar issues regarding the ability to export a folder as a single PDF/etc. and they all redirect here with no useful information or updates for Typora. Our team was looking into using Typora for our projects documentation, but without this feature we're unable to use it as cramming everything into a single .md is not possible.

**Comment by TheFern2** [comment]
> > Is there anything being done on this 7 year old issue? It appears that multiple people have opened similar issues regarding the ability to export a folder as a single PDF/etc. and they all redirect here with no useful information or updates for Typora. Our team was looking into using Typora for our projects documentation, but without this feature we're unable to use it as cramming everything into a single .md is not possible.
> 
> 
> 
> I love Typora too much for the live-preview and editing so what I've done is use the m4 macro processor (https://gnu.org/software/m4/manual/) to put together my "chapters" into one big md file and then I use Typora to open that and print/save it as the format I need.  So my workflow is 
> 
> 
> 
> 1. write chapters or sections as .md files in Typora
> 
> 2. create a .m4 file to explain how to combine them
> 
> 3. use m4 to create a new concatenated md file
> 
> 4. use Typora on the new .md file
> 
> 
> 
> The hassle is worth it to me. I also reuse chapters and sections in different reports, so many times all I need to do is create a new .m4 file and I'm 90% of the way done.  

Nice! This is exactly what I'm talking about just use whatever editor you like and find a printing solution both don't have to be the same. Too many options out there to just be waiting patiently lol

**Comment by mikedeskevich** [comment]
> I don't think you have to do anything special to concatenate several txt/md files into one file for further processing? Most UNIX / macOS have the 'cat' command e.g. cat file1.md file2.md file3.md > file4.md 

Yes, .md is just plain old text and catting them is fine.  I do a little more with my .m4 templates to manage the layout for different kinds of reports - e.g., insert a common introduction, table of contents, appendices.  It's probably 95% "include()", but there is other formatting there.  

I probably should use something like quarto for my work, but the people I work with would never want to learn that, and the live preview of Typroa helps my brain so much that I write much faster, so that's why I do this extra little workflow.

**Comment by Ziflin** [comment]
@TheFern2 - We use quite a few of Typora's extended markdown features for math blocks/inline math equations, customized css, etc. and need the PDF/HTML output to match what we see in Typora -- which so far it does. Using some other app and toolset is just not something we're interested in doing/maintaining, but that's great if it works well for you. 

@shoesforindustry @mikedeskevich I don't believe just concatenating .md files works if you have links between files which we do.

**Comment by TheFern2** [comment]
> @TheFern2 - We use quite a few of Typora's extended markdown features for math blocks/inline math equations, customized css, etc. and need the PDF/HTML output to match what we see in Typora -- which so far it does. Using some other app and toolset is just not something we're interested in doing/maintaining, but that's great if it works well for you. 
> 
> 
> 
> @shoesforindustry @mikedeskevich I don't believe just concatenating .md files works if you have links between files which we do.

Everything you like about typora has nothing to do with typora is latex which you can manipulate with pandoc as well. And with pandoc you can do links as well, headers, margins, etc. I wouldn't be surprised if typora uses pandoc under the hood as well but obviously not very well.

**Comment by shoesforindustry** [comment]
> @TheFern2 I don't believe just concatenating .md files works if you have links between files which we do.

Links of course will be copied, but links between files will be just that, linking to the original file. I presume you want links between the sections of original newly concatenated document? 

I guess it depend on how you do the link in the originals? It can work, say if you link to headings with anchor style links:

In file1.md
```
# File 1
A link to anchor (heading) in [File 2](#file-2)
```

and in file2.md a heading
`# File 2`

This still works when you concatenate the files with 
`cat file1.md file2.md > file3.md`

In file3.md
```
# File 1

A link to anchor (heading) in [File 2](#file-2)

#File 2
```
Of course your needs maybe more complex than this and there are other suggestions here that could work for you. We edit in Typora and then compile with Pandoc. This way you can do pretty much what you want. If you use a YAML header in your markdown for the compile instructions/metadata you can keep it relatively self contained.

> @mikedeskevich There are I guess many ways to do this, we each have our own needs. I was just trying to show a simple solution but your looks good to, I shall have a closer look at m4 :)
